### PR TITLE
feat(core): middleware introduced

### PR DESCRIPTION
**Pull request by Guikingone**

### CODE REVIEW

**Comment by Guikingone** [comment]
@jmsche Do you have time to review it? 

Thanks 🙏🏻

File: .github/workflows/daily.yml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-name: SchedulerBundle - Daily
+name: Daily
```
Comment (jmsche) [inline .github/workflows/daily.yml:1]: Requires to change badge's URL on README

File: doc/middleware.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
+# Middleware
+
+- [Scheduling](middleware.md#Scheduling)
+- [Execution](middleware.md#Execution)
+- [Order](middleware.md#Order)
```
Comment (jmsche) [inline doc/middleware.md:5]:
```suggestion
- [Scheduling](#scheduling)
- [Execution](#execution)
- [Order](#order)
```

File: doc/middleware.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+This bundle defines middleware related to execution and scheduling phases.
+
+Middlewares are "man in the middle" that allows you to interact with the task
+that is about to be scheduled/executed or even after the scheduling/execution.
+
+There's two type of middleware:
```
Comment (jmsche) [inline doc/middleware.md:12]:
```suggestion
There are two types of middleware:
```

File: doc/middleware.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    {
+    }
+
+    public function postScheduling(TaskInterface $task, SchedulerInterface $scheduler) : void
+    {
+    }
+}
+```
+
+Both method receives the current task (before scheduling it and sending it through transport) along with the scheduler.
```
Comment (jmsche) [inline doc/middleware.md:54]:
```suggestion
Both methods receive the current task (before scheduling it and sending it through transport) along with the scheduler.
```

File: doc/middleware.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    {
+    }
+
+    public function postExecute(TaskInterface $task) : void
+    {
+    }
+}
+```
+
+Both method receives the current task.
```
Comment (jmsche) [inline doc/middleware.md:90]:
```suggestion
Both methods receive the current task.
```

File: doc/middleware.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+### Extra informations
```
Comment (jmsche) [inline doc/middleware.md:92]:
```suggestion
### Extra information
```

File: doc/middleware.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+- Implementing both interfaces for each middleware is not required, your middleware can be focused on a single one.
+- A middleware can interact during both process by implementing the desired interfaces, 
+  each stack sort related middleware before interacting with it.
```
Comment (jmsche) [inline doc/middleware.md:96]:
```suggestion
- A middleware can interact during both processes by implementing the desired interfaces, 
  each stack sorts related middlewares before interacting with them.
```

File: doc/test.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+| ----------------------------------------------------------------| -------------------------------------|
+| [`TaskExecuted`](../src/Test/Constraint/TaskExecuted.php)       | Assert on the executed task count    |
+| [`TaskFailed`](../src/Test/Constraint/TaskFailed.php)           | Assert on the failed task count      |
+| [`TaskQueued`](../src/Test/Constraint/TaskQueued.php)           | Assert on the queued task count      |
+| [`TaskScheduled`](../src/Test/Constraint/TaskScheduled.php)     | Assert on the scheduled task count   |
+| [`TaskUnscheduled`](../src/Test/Constraint/TaskUnscheduled.php) | Assert on the unscheduled task count |
+
+## Tests trait
+
+A little bonus is the `SchedulerAssertionTrait` that can be used on tests that extends `KernelTestCase`:
```
Comment (jmsche) [inline doc/test.md:15]:
```suggestion
A little bonus is the `SchedulerAssertionTrait` that can be used on tests that extend `KernelTestCase`:
```

File: phpunit.xml.dist [hunk 9]
```diff
@@ -1,30 +1,31 @@
 <?xml version="1.0" encoding="UTF-8"?>
-
 <phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-         xsi:noNamespaceSchemaLocation="http://schema.phpunit.de/5.2/phpunit.xsd"
+         xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd"
```
Comment (jmsche) [inline phpunit.xml.dist:3]:
```suggestion
         xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.5/phpunit.xsd"
```
9.5 is specified in composer.json

File: src/Middleware/PostSchedulingMiddlewareInterface.php [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+use Throwable;
+
+/**
+ * @author Guillaume Loulier <contact@guillaumeloulier.fr>
+ */
+interface PostSchedulingMiddlewareInterface
+{
+    /**
+     * Allow to execute logic after scheduling the task
+     * (the @param TaskInterface $task is the one passed through {@see SchedulerInterface::schedule()} and AFTER the transport store it).
```
Comment (jmsche) [inline src/Middleware/PostSchedulingMiddlewareInterface.php:18]:
```suggestion
     * (the @param TaskInterface $task is the one passed through {@see SchedulerInterface::schedule()} and AFTER the transport stores it).
```

File: src/Middleware/RateLimiterMiddleware.php [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            return;
+        }
+
+        $limiter = $this->rateLimiter->create($task->getName());
+
+        try {
+            $limiter->reserve($task->getMaxExecution());
+        } catch (ReserveNotSupportedException $exception) {
+            $this->logger->critical(sprintf(
+                'A reservation cannot be created for task "%s", please ensure that the policy used support it',
```
Comment (jmsche) [inline src/Middleware/RateLimiterMiddleware.php:49]:
```suggestion
                'A reservation cannot be created for task "%s", please ensure that the policy used supports it.',
```

File: src/Middleware/TaskCallbackMiddleware.php [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+     * {@inheritdoc}
+     */
+    public function preScheduling(TaskInterface $task, SchedulerInterface $scheduler): void
+    {
+        if (is_null($task->getBeforeScheduling())) {
+            return;
+        }
+
+        if (false === call_user_func($task->getBeforeScheduling(), $task)) {
+            throw new MiddlewareException('The task cannot be scheduled as en error occurred on the before scheduling callback');
```
Comment (jmsche) [inline src/Middleware/TaskCallbackMiddleware.php:30]:
```suggestion
            throw new MiddlewareException('The task cannot be scheduled as an error occurred on the before scheduling callback.');
```

File: src/Middleware/TaskCallbackMiddleware.php [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    public function postScheduling(TaskInterface $task, SchedulerInterface $scheduler): void
+    {
+        if (is_null($task->getAfterScheduling())) {
+            return;
+        }
+
+        if (false === call_user_func($task->getAfterScheduling(), $task)) {
+            $scheduler->unschedule($task->getName());
+
+            throw new MiddlewareException('The task has encounter an error after scheduling, it has been unscheduled');
```
Comment (jmsche) [inline src/Middleware/TaskCallbackMiddleware.php:46]:
```suggestion
            throw new MiddlewareException('The task has encountered an error after scheduling, it has been unscheduled.');
```

File: src/Middleware/TaskCallbackMiddleware.php [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+
+    public function preExecute(TaskInterface $task): void
+    {
+        if (is_null($task->getBeforeExecuting())) {
+            return;
+        }
+
+        if (false === call_user_func($task->getBeforeExecuting(), $task)) {
+            throw new MiddlewareException(sprintf(
+                'The task "%s" has encounter an error when executing the %s::getBeforeExecuting() callback',
```
Comment (jmsche) [inline src/Middleware/TaskCallbackMiddleware.php:58]:
```suggestion
                'The task "%s" has encountered an error when executing the %s::getBeforeExecuting() callback.',
```

File: src/Middleware/TaskCallbackMiddleware.php [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+    public function postExecute(TaskInterface $task): void
+    {
+        if (is_null($task->getAfterExecuting())) {
+            return;
+        }
+
+        if (false === call_user_func($task->getAfterExecuting(), $task)) {
+            throw new MiddlewareException(sprintf(
+                'The task "%s" has encounter an error when executing the %s::getAfterExecuting() callback',
```
Comment (jmsche) [inline src/Middleware/TaskCallbackMiddleware.php:73]:
```suggestion
                'The task "%s" has encountered an error when executing the %s::getAfterExecuting() callback.',
```

File: src/Resources/views/Collector/data_collector.html.twig [hunk 16]
```diff
@@ -3,7 +3,7 @@
 {% import _self as helper %}
 
 {% block toolbar %}
-    {% set events = collector.events %}
+    {% set events = collector.getEvents() %}
```
Comment (jmsche) [inline src/Resources/views/Collector/data_collector.html.twig:6]: Any existing conflict?

File: src/Task/AbstractTask.php [hunk 17]
```diff
@@ -66,6 +66,8 @@ protected function defineOptions(array $options = [], array $additionalOptions =
             'execution_end_time' => null,
             'last_execution' => null,
             'max_duration' => null,
+            'max_execution' => null,
+            'max_retry' => null,
```
Comment (jmsche) [inline src/Task/AbstractTask.php:70]: What about using plural? (max_executions & max_retries)

Also applies to getters, setters & $resolver->setInfo()

File: src/Task/TaskInterface.php [hunk 18]
```diff
@@ -186,6 +186,14 @@ public function setMaxDuration(float $maxDuration = null): self;
 
     public function getMaxDuration(): ?float;
 
+    public function setMaxExecution(int $maxExecution = null): TaskInterface;
+
+    public function getMaxExecution(): ?int;
+
+    public function setMaxRetry(int $maxRetry = null): TaskInterface;
+
+    public function getMaxRetry(): ?int;
```
Comment (jmsche) [inline src/Task/TaskInterface.php:195]: Same comment about plural as AbstractTask

File: tests/EventListener/TaskSubscriberTest.php [hunk 19]
```diff
@@ -73,6 +74,8 @@ public function testValidPathCannotBeHandledWithoutParams(): void
         $subscriber = new TaskSubscriber($scheduler, $worker, $eventSubscriber, $serializer);
 
         self::expectException(InvalidArgumentException::class);
+        self::expectExceptionMessage('A GET request should at least contains a task name or its expression!');
```
Comment (jmsche) [inline tests/EventListener/TaskSubscriberTest.php:77]:
```suggestion
        self::expectExceptionMessage('A GET request should at least contain a task name or its expression!');
```

**Review by jmsche — CHANGES_REQUESTED** [review]
I guess applying suggested changes like reworded exception messages will require to adapt tests

File: tests/EventListener/TaskSubscriberTest.php (hunk 19 above)
Comment (Guikingone) [reply tests/EventListener/TaskSubscriberTest.php:75]: Fixed

File: src/Middleware/TaskCallbackMiddleware.php (hunk 15 above)
Comment (Guikingone) [reply src/Middleware/TaskCallbackMiddleware.php:73]: Fixed

File: doc/middleware.md (hunk 2 above)
Comment (Guikingone) [reply doc/middleware.md:5]: Fixed

File: doc/middleware.md (hunk 3 above)
Comment (Guikingone) [reply doc/middleware.md:12]: Fixed

File: doc/middleware.md (hunk 5 above)
Comment (Guikingone) [reply doc/middleware.md:90]: Fixed

File: doc/middleware.md (hunk 4 above)
Comment (Guikingone) [reply doc/middleware.md:54]: Fixed

File: doc/middleware.md (hunk 6 above)
Comment (Guikingone) [reply doc/middleware.md:92]: Fixed

File: doc/test.md (hunk 8 above)
Comment (Guikingone) [reply doc/test.md:15]: Fixed

File: phpunit.xml.dist (hunk 9 above)
Comment (Guikingone) [reply phpunit.xml.dist:3]: Fixed

File: src/Middleware/PostSchedulingMiddlewareInterface.php (hunk 10 above)
Comment (Guikingone) [reply src/Middleware/PostSchedulingMiddlewareInterface.php:18]: Fixed

File: src/Middleware/TaskCallbackMiddleware.php (hunk 13 above)
Comment (Guikingone) [reply src/Middleware/TaskCallbackMiddleware.php]: Fixed

File: .github/workflows/daily.yml (hunk 1 above)
Comment (Guikingone) [reply .github/workflows/daily.yml:1]: Fixed

File: doc/middleware.md (hunk 7 above)
Comment (Guikingone) [reply doc/middleware.md:96]: Fixed

File: src/Middleware/RateLimiterMiddleware.php (hunk 11 above)
Comment (Guikingone) [reply src/Middleware/RateLimiterMiddleware.php:49]: Fixed

File: src/Middleware/TaskCallbackMiddleware.php (hunk 12 above)
Comment (Guikingone) [reply src/Middleware/TaskCallbackMiddleware.php:30]: Fixed

File: src/Middleware/TaskCallbackMiddleware.php (hunk 14 above)
Comment (Guikingone) [reply src/Middleware/TaskCallbackMiddleware.php:58]: Fixed

File: src/Resources/views/Collector/data_collector.html.twig (hunk 16 above)
Comment (Guikingone) [reply src/Resources/views/Collector/data_collector.html.twig:6]: Not a single one, just a better debug experience IMHO

File: src/Task/AbstractTask.php (hunk 17 above)
Comment (Guikingone) [reply src/Task/AbstractTask.php:70]: Agree, improved

File: src/Task/TaskInterface.php (hunk 18 above)
Comment (Guikingone) [reply src/Task/TaskInterface.php:211]: Fixed

**Pull request merged**